src/Microsoft.Azure.Batch.SoftwareEntitlement.Server/Controllers/SoftwareEntitlementsController.cs

```diff
@@ -84,12 +90,36 @@ public SoftwareEntitlementsController(ServerOptions serverOptions, ILogger logge
                 var remoteAddress = HttpContext.Connection.RemoteIpAddress;
                 _logger.LogDebug("Remote Address: {Address}", remoteAddress);
 
+                if (entitlementRequest.HostId == null && ApiRequiresHostId(apiVersion))
+                {
+                    return CreateMissingHostIdError();
+                }
+
+                if (!entitlementRequest.Cores.HasValue && ApiRequiresCpuCoreCount(apiVersion))
+                {
+                    return CreateMissingCpuCoreCountError();
+                }
+
                 Errorable<NodeEntitlements> verificationResult = _verifier.Verify(
                     entitlementRequest.Token,
                     _serverOptions.Audience,
                     _serverOptions.Issuer,
                     entitlementRequest.ApplicationId,
-                    remoteAddress);
+                    remoteAddress,
+                    entitlementRequest.Cores);
+
+                // Check the host ID from the request against the entitlement ID we've just
+                // extracted from the token.
+                verificationResult = verificationResult.Then(entitlement =>
+                {
+                    var allowed =
+                        !ApiRequiresHostId(apiVersion) ||
```

Comment:
We normally put `||` and `&&` on the start of lines (there's evidence that punctuation at the ends of lines is much easier for people to not read).

I argued that `=` and `=>` should be on the starts of lines too, but loss that argument. 😀